designs/2019-eslintrc-improvements/README.md

```diff
@@ -0,0 +1,358 @@
+
+## Backwards Compatibility Analysis
+
+Most cases work fine as is.
+But I guess there are some breaking changes.
+
+### ✅ CLIEngine#addPlugin
+
+It can work fine as is.
+`CLIEngine` has a `Map<string, Plugin>` for the added plugin, and it gives `FileEnumerator` (then `ConfigArrayFactory.load*` methods) the map, and the loading logic uses the map to load plugins.
+
+### ✅ Linter
+
+It can work fine as is.
+
+Only if a given `config` has `extractConfig` method, `Linter` does the setup process for the config. Otherwise, it works the same as currently.
+
+### ✅ LintResultCache
+
+It can work fine as is.
+
+The `ConfigArray` has all information to distinguish if the config was changed since the last time.
+
+### ⚠️ `overrides` affects target files.
+
+For example, `files: ["*.ts"]` in configuration makes ESLint checking TypeScript files as well.
+This is a breaking change, but I think reasonable. If we don't want this change, we can remove this feature from this RFC.
+
+### ⚠️ Fix a surprised behavior of `overrides`.
+
+Currently, `overrides` is applied after all `overrides` properties are merged. This means that the `overrides` in a shareable config priors to the setting in your `.eslintrc`.
+
+```yml
+extends:
+    - foo
+rules:
+    eqeqeq: error # silently ignored if the "foo" has the "eqeqeq" setting in "overrides".
+```
+
+After this proposal, the setting in your `.eslintrc` priors to the setting in shareable configs always.
+This is a breaking change, but I think reasonable.
+
+### ⚠️ It looks plugins up from the location where the config file is.
+
+This means we get the consistent way to load dependencies on `extends`, `parser`, and `plugins`.
+
+Currently, ESLint looks shareable configs and parsers relatively from the location where the config file is. But ESLint looks plugins from the project root. This is forcing inconvenient to shareable config maintainers ([eslint/eslint#3458], [eslint/eslint#10643]).
+
+This proposal makes those behaviors consistent.
+
+Each element of `ConfigArray` has the loaded plugins. When ESLint merges those by `ConfigArray#extractConfig(filePath)`, if one same plugin loaded from different config files (except self-loading), it throws a confliction error because of https://gist.github.com/not-an-aardvark/169bede8072c31a500e018ed7d6a8915.
+
+This behavior is described on [#14] as well. This change has huge impact, probably we should discuss on [#14] individually.
+
+People can resolve this problem with `.eslintrc.js` and to install needed plugins:
+
+```js
+const foo = require("eslint-config-foo")
+const bar = require("eslint-config-bar")
+
+module.exports = {
+    overrides: [
+        Object.assign({ files: "*.js" }, foo),
+        Object.assign({ files: "*.js" }, bar),
```

Comment:
I don't think this would work in general because if `eslint-config-foo` contains `extends` clauses, they are only valid when resolved relative to the location of `eslint-config-foo`. (This is a disadvantage of using strings with `extends`.)

Maybe we could provide a utility function that rewrites all paths in a config to absolute paths.

Reply:
I got it.

I'm considering either of two options.

1. Provides a utility to rewrite relative paths to absolute paths.
2. Allows the same plugins between a config file and extended configs.